atomix/cluster/src/main/java/io/atomix/raft/impl/DefaultRaftServer.java

```diff
@@ -123,22 +124,27 @@ public void removeFailureListener(final Runnable failureListener) {
    *
    * @return A completable future to be completed once the server has been shutdown.
    */
+  @Override
   public CompletableFuture<Void> shutdown() {
-    if (!started) {
+    if (!started && !stopped) {
       return Futures.exceptionalFuture(new IllegalStateException("Server not running"));
     }
 
+    if (stopped) {
```

Comment:
@korthout do you think it makes sense to replace the boolean flags with just having a future for starting and stopping?

If starting future is null, you know it hasn't been started etc. If you call stop twice you just return the stopping future. I think it is not necessary to return an exceptionally future, for me it is not an error to close something what haven't been started. What do you think?

Reply:
It makes sense but since you can start, stop and leave async this gets very complex. It requires keeping track of a queue of futures (or something similar to ZeebePartitionTransition). It would be easier if we could just remove leaving behaviour, since we don't use this anyways. I'll create an issue for it.

Reply:
Created here: https://github.com/zeebe-io/zeebe/issues/5663